src/Core/CustomRedirects/CustomRedirectCollection.cs

```diff
@@ -71,6 +80,14 @@ private CustomRedirect FindInternal(string url)
                 return redirect;
             }
 
+            // working with local copy to avoid multi-threading issues
+            var redirectsZA = _redirectsZACache;
+            if (redirectsZA == null)
+            {
+                redirectsZA = _quickLookupTable.OrderByDescending(x => x.Key, StringComparer.OrdinalIgnoreCase).ToArray();
```

Comment:
What performance decrease this will make? Could you compare the old version with the new one with 10k, 100k, redirects?

Reply:
in original code, we also iterate array - so no differences here
the only case in this new code when we create new array is when we manually added redirects from UI, so I would say we can completely ignore that difference.

Reply:
The difference is that we are sorting Z to A here and with a large number of items it might take some time.

Reply:
Hi @marisks, I would say this is not something we should care about. On a live site (assuming everything is configured and we don't touch redirect gadget), sorting is being executed just _once_, during module initialization. 

In the same time, if we don't have exact match (like spamming site with fake URLs), for **every** 404 status we will be traversing whole redirects array several times - we call `FindInternal` three times here https://github.com/Geta/404handler/blob/ce5ce1e9bd65743c5e44c46c80d24a8ad3bfff83/src/Core/CustomRedirects/CustomRedirectCollection.cs#L32 and then iterate through whole collection here https://github.com/Geta/404handler/blob/ce5ce1e9bd65743c5e44c46c80d24a8ad3bfff83/src/Core/CustomRedirects/CustomRedirectCollection.cs#L84

I was playing with a bit better condition:
1. we have only relative URLs defined, and 
2. I was requesting exactly matched URI

Due to the code written here https://github.com/Geta/404handler/blob/ce5ce1e9bd65743c5e44c46c80d24a8ad3bfff83/src/Core/CustomRedirects/CustomRedirectCollection.cs#L35 it always first tries to check absolute URL first, but will never find it with "all relative redirects defined" condition (I would say, pretty common case in a wild), so we will have exactly one go through whole redirects list anyway.

On my local machine I have figures like these (release build, console app, single thread):

|approx # of redirects|Time to sort, s|AVG redirects handled per second|
---:|---:|---
500 | 0,0009 | 8 000
5 000 | 0,017 | 700
50 000 | 0,2 | 67
500 000 | 4 | 5

As you see, tool itself is not suitable for high numbers of redirects - for 50k redirects defined sort time is still acceptable, but tool behavior is not, really - handling just 67 _exactly matched_ redirects per second with whole CPU core used.

Having all these, I would say additional sort operation (again - once per app domain, in average) won't make things worse, while it's really small change which most probably doesn't break the logic we have so far.

(as a side note - current performance is really poor, but changing that would require quite a rewriting, definitely in separate PR; and that most probably would be a breaking change - as I see how redirects becomes working in _more correct_ way, but that will be _different_ way)

Reply:
If `CustomRedirectCollection` is "singleton" then yes, sorting will not affect it much - just on the first call. Just can't see the whole context in GitHub, so asking questions :)

Regarding the order of lookup (absolute URL first) - it is how the library worked before. Changing the order might break some clients. Actually, your change will break some too, but for good :) This version then should be a major update.

Reply:
Hi @marisks I would say yes, it supposed to be singleton. 

`CustomRedirectCollection` is being used as part of `CustomRedirectHandler`, which is available through static `Current` property only: https://github.com/Geta/404handler/blob/6f7f0fa17ed07fcabd54c6e0505a8f6b65a748a9/src/Core/CustomRedirects/CustomRedirectHandler.cs#L57

It is being stored in episerver cache, so whenever collection is being changed from gadget UI, remote event is sent to all instances in load balancer to remove that from cache:

https://github.com/Geta/404handler/blob/6f7f0fa17ed07fcabd54c6e0505a8f6b65a748a9/src/Core/CustomRedirects/CustomRedirectHandler.cs#L113

(in the same time, as it is a shared instance, we should be careful with multi-thread safety)